### PR TITLE
Skwasm single threaded

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This PR creates a single-threaded version of the skwasm renderer, appropriate for non-crossOriginIsolated browsing contexts.
* The single threaded renderer is essentially the same as the multi-threaded renderer, except instead of spawning a web worker and posting messages to it, it simply schedules microtasks on the main thread in their place.
* The new renderer is vended as `skwasm_st.js` and `skwasm_st.wasm` in the same location as multithreaded skwasm. In order to properly build and function, we needed some fixes I put into emscripten that landed in version 3.1.70. That version also changed some behavior that required a few fixes to the CanvasKit build files.
* The skwasm loader in flutter.js has been modified to use the skwasm_st variants when encountering a non-crossOriginIsolated context but a browser and configuration that otherwise would allow the use of skwasm. I also added a new `forceSingleThreadedSkwasm` option to the flutter configuration so that we can override this behavior, especially so that we can accurately benchmark the single threaded renderer in a crossOriginIsolated environment.
* I also consolidated a bunch of our shards that run tests to just have one per browser/platform combination, so four total. This will address https://github.com/flutter/flutter/issues/124682

### CODE REVIEW

File: lib/web_ui/dev/steps/run_suite_step.dart [hunk 1]
```diff
@@ -192,17 +194,26 @@ class RunSuiteStep implements PipelineStep {
...
-        'Browser': suite.runConfig.browser.name,
-        if (isWasm) 'Wasm': 'true',
-        'Renderer': renderer.name,
-        if (variant != null) 'CanvasKitVariant': variant.name,
-      },
+      dimensions: dimensions,
     );
 
     if (await _checkSkiaClient(skiaClient)) {
+      print('Successfully checked Skia Gold Client');
```
Comment (harryterkelsen) [inline lib/web_ui/dev/steps/run_suite_step.dart:216]: Are these leftover debug prints or are they intended to be part of the output (e.g. to help debug flaky tests on CI)

File: lib/web_ui/dev/generate_builder_json.dart [hunk 2]
```diff
@@ -110,71 +110,62 @@ Map<String, dynamic> _getBundleBuildStep(TestBundle bundle) {
...
-            ],
-            'script': 'flutter/lib/web_ui/dev/felt',
-          }
-        ]
-      }
-    );
+      if (browser == BrowserName.firefox)
+        <String, dynamic>{
+          'dependency': 'firefox',
+          'version': 'version:106.0',
```
Comment (harryterkelsen) [inline lib/web_ui/dev/generate_builder_json.dart:158]: I find it a little scary to hard code dependency versions in code. Is it documented that when we upgrade these dependencies that we should also change the code on these lines?

**Review by harryterkelsen — APPROVED** [review]
LGTM with a couple questions

File: lib/web_ui/dev/generate_builder_json.dart (hunk 2 above)
Comment (eyebrowsoffire) [reply lib/web_ui/dev/generate_builder_json.dart:158]: Yeah actually I agree. Maybe I can parse these out of the `browser_lock.yaml` file or whatever.

File: lib/web_ui/dev/steps/run_suite_step.dart (hunk 1 above)
Comment (eyebrowsoffire) [reply lib/web_ui/dev/steps/run_suite_step.dart:216]: I found them actually kind of useful enough to leave in. They only are printed once for each suite when they run, so they aren't particularly chatty. If you don't like them in here I can remove them, but it's actually nice to be able to make sure we created a skia gold client with the right dimensions.

Comment (harryterkelsen) [reply lib/web_ui/dev/steps/run_suite_step.dart:216]: Seems fine to keep in, then.

**Comment by jonahwilliams** [comment]
reason for revert: failing to roll into framework: https://github.com/flutter/flutter/pull/157919

**Comment by brian-superlist** [comment]
Heya, very cool work! I was trying this out, but ran into the following error and the web app doesn't load for me.

What I tried:

- Upgrading to Flutter/master with this engine roll in place
- Build our web app, adding the `forceSingleThreadedSkwasm` to the `_flutter.loader.load({config: {forceSingleThreadedSkwasm: true}` to ensure we're running in single threaded mode. Command used: `flutter build web --release --csp --source-maps --target lib/main_dev.dart --wasm --no-strip-wasm`
- Ran the web app with dhttpd: `dart pub global run dhttpd --path build/web`
- Open localhost:8080 in Chrome Version 130.0.6723.70 (Official Build) (arm64)
- I can confirm that it's loading the wasm files (main.dart.wasm in the sources panel of the Chrome Dev Tools)
- I see it also loading the new skwasm_st.js from https://www.gstatic.com/flutter-canvaskit/05cb5d7f7939da58ef9897997004f181f2699075/skwasm_st.js

The following error is thrown:

```
Dart stack trace:
    at module0.initializeEngineServices closure initializeRendererCallback at org-dartlang-sdk:///lib/_engine/engine/initialization.dart:196:43 inner (initialization.dart:196:43)
    at module0.initializeEngineServices inner (initialization.dart:196:43)
    at module0.bootstrapEngine closure at org-dartlang-sdk:///lib/ui_web/ui_web/initialization.dart:33:23 inner (initialization.dart:110:14)
    at module0.bootstrapEngine closure at org-dartlang-sdk:///lib/ui_web/ui_web/initialization.dart:33:23 (initialization.dart:33:23)
    at module0.closure wrapper at org-dartlang-sdk:///lib/ui_web/ui_web/initialization.dart:33:23 trampoline_118538 (main.dart.wasm:0x979589)
    at module0.AppBootstrap.prepareEngineInitializer closure at org-dartlang-sdk:///lib/_engine/engine/app_bootstrap.dart:47:25 inner (app_bootstrap.dart:48:32)
    at module0.AppBootstrap.prepareEngineInitializer closure at org-dartlang-sdk:///lib/_engine/engine/app_bootstrap.dart:47:25 (app_bootstrap.dart:47:25)
    at module0.closure wrapper at org-dartlang-sdk:///lib/_engine/engine/app_bootstrap.dart:47:25 trampoline_118554 (main.dart.wasm:0x97968c)
    at module0.initializeEngineServices closure initializeRendererCallback at org-dartlang-sdk:///lib/_engine/engine/initialization.dart:196:43 inner (initialization.dart:196:43)
    at module0.initializeEngineServices inner (initialization.dart:196:43)
    at module0.bootstrapEngine closure at org-dartlang-sdk:///lib/ui_web/ui_web/initialization.dart:33:23 inner (initialization.dart:110:14)
    at module0.bootstrapEngine closure at org-dartlang-sdk:///lib/ui_web/ui_web/initialization.dart:33:23 (initialization.dart:33:23)
    at module0.closure wrapper at org-dartlang-sdk:///lib/ui_web/ui_web/initialization.dart:33:23 trampoline_118538 (main.dart.wasm:0x979589)
    at module0.AppBootstrap.prepareEngineInitializer closure at org-dartlang-sdk:///lib/_engine/engine/app_bootstrap.dart:47:25 inner (app_bootstrap.dart:48:32)
    at module0.AppBootstrap.prepareEngineInitializer closure at org-dartlang-sdk:///lib/_engine/engine/app_bootstrap.dart:47:25 (app_bootstrap.dart:47:25)
    at module0.closure wrapper at org-dartlang-sdk:///lib/_engine/engine/app_bootstrap.dart:47:25 trampoline_118554 (main.dart.wasm:0x97968c)
    at _2868 (main.dart.mjs:1286:16)
    at module0.JSFunctionUnsafeUtilExtension|callAsConstructor (js_interop_unsafe_patch.dart:61:22)
    at module0.CustomFutureOfJSAnyToJSPromise|get#toPromise closure at org-dartlang-sdk:///lib/_engine/engine/js_interop/js_promise.dart:18:19 (js_promise.dart:27:30)
    at module0.closure wrapper at org-dartlang-sdk:///lib/_engine/engine/js_interop/js_promise.dart:18:19 trampoline (main.dart.wasm:0x979d49)
    at module0._RootZone.runBinary (zone.dart:1681:54)
    at module0._Future._propagateToListeners (future_impl.dart:219:22)
    at module0._Future._completeError (future_impl.dart:711:5)
    at module0._Future._asyncCompleteError closure at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:796:29 (future_impl.dart:797:7)
    at module0.closure wrapper at org-dartlang-sdk:///dart-sdk/lib/async/future_impl.dart:796:29 trampoline (main.dart.wasm:0x38b771)
    at module0._startMicrotaskLoop (schedule_microtask.dart:40:21)
```


I've also sent an email with some of this info to @kevmoo since he's been coordinating with us to get our app running with WASM, but I thought I'd put more info here.

I'm excited to get WASM going! So far, we've seen some big performance improvements with the multi-threaded renderer, but we have some use cases which don't work with the restrictive COEP & COOP headers. Thanks for all the effort.

**Comment by brian-superlist** [comment]
Ah wait, my bad -- I looked at the project code for a fresh project vs our custom code we have. We had actually enabled a couple of performance optimizations for firefox with JS, but that was a bad idea with the new wasm renderer. 

Apologies for the noise and thanks again for the contribution. Here are the offending lines I had to remove:

```
window.createImageBitmap = null;
window.OffscreenCanvas = null;
```

**Comment by eyebrowsoffire** [comment]
@brian-superlist  Ah good to hear. Yes the skwasm web renderer requires those APIs so that makes sense. Glad to hear it's working!

**Pull request merged**